Add PrusaLink nozzle and mmu support

deimosfr opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding another endpoint to PrusaLink to support MMU enabled/disabled and nozzle size.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33878

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

deimosfr committed:
feat: add prusa nozzle and mmu support

agners reviewed (changes requested):
On my Prusa MK3.9 this PR leads to the following error:

```
2024-06-25 20:09:09.521 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up prusalink platform for sensor
Traceback (most recent call last):
  File "/home/sag/projects/home-assistant/core/homeassistant/helpers/entity_platform.py", line 364, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/home/sag/projects/home-assistant/core/homeassistant/components/prusalink/sensor.py", line 222, in async_setup_entry
    coordinator = coordinators[coordinator_type]
                  ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
KeyError: 'info'
```

This requires another coordinate to handle the `ServerInfo` endpoint.

agners committed:
Add InfoUpdateCoordinator and Binary sensor support

Both, the MMU and Nozzle Diameter require the PrinterInfo. This adds
an update coordinator for the PrinterInfo.

Furthermore, the MMU requires a binary sensor to indicate if the MMU is
enabled or not. This adds a binary sensor platform support to the
prusalink integration to reflect the MMU state.

agners reviewed (approved):
So I went ahead and implemented the changes. It turned out to be a bit bigger than I expected, since the MMU property is a boolean: The prusalink integration did not yet support the `binary_sensor` platform :cold_sweat: . Now it does, and everything should be ready. I've tested this with a Prusa MK3.9.

However, this needs a second review, since most of the change is now from me :innocent:

frenck reviewed (approved):
Looks good to me! Thanks both 👍 

../Frenck

deimosfr commented:
Wow thanks a lot!

MartinHjelmare commented on tests/components/prusalink/test_binary_sensor.py:
```diff
@@ -0,0 +1,33 @@
+    with (
+        patch("homeassistant.components.prusalink.PLATFORMS", [Platform.BINARY_SENSOR]),
+        patch(
+            "homeassistant.helpers.entity.Entity.entity_registry_enabled_default",
```
We have a fixture for this already.

https://github.com/home-assistant/core/blob/ff9937f942828c18c82e8e6ff66ca9f0004fcea8/tests/components/conftest.py#L45-L52

MartinHjelmare reviewed (commented):
Please address the comment in a new PR. Thanks!
